### ISSUE TITLE
Need a system for ensuring minimum (native) dependencies

### ISSUE THREAD

**Issue by bartonjs** [dotnet/cli#674]
CoreFX has dependencies on native libraries for functionality we haven't duplicated.  Many of these dependencies may already met on standard installations of our intended platforms, but at least one (OpenSSL \> 1.0.0 on OS X) is not.

* ICU (libicucore) - System.Globalization.Native (coreclr)
* zlib (libz) - System.IO.Compression.Native (corefx)
* curl (libcurl) - System.Net.Http.Native (corefx)
* OpenSSL LTS (libcrypto, libssl) - System.Security.Cryptography.Native (corefx)
    * OS X includes libssl and libcrypto on-box, but they are version 0.9.8 which goes out of LTS this week.

Related Issues:
* https://github.com/dotnet/cli/issues/77#issuecomment-158512797 (mentions ICU dependency)
* https://github.com/dotnet/corefx/issues/4737

For package installations of the cli (brew, dpkg, rpm, etc) these should be encoded as dependencies.

**Comment by bartonjs** [comment]
FYI: @piotrpMSFT, @ellismg 

(Matt, I feel like the ICU dependencies are tighter than that, we need specific versions that correspond with whatever version we built System.Globalization.Native against, right?)

**Comment by ellismg** [comment]
Regarding ICU, correct.  Most platforms do not disable function renaming and so we link against exports that have version numbers baked into them.

**Comment by piotrpMSFT** [comment]
To be clear, these are dependencies both for the CLI as well as any app produced atop CoreFx, right? The idea of a self-contained app folder is unrealistic and dependency validation is necessary?

**Comment by davidfowl** [comment]
I was going to mention that's this is blurring runtime for the CLI itself and the applications produced by the CLI toolchain

**Comment by bartonjs** [comment]
The dependencies are of CoreFX, so primarily my concern is about apps produced by the toolchain.  But since the CLI is itself a CoreFX application, it applies to both.

The self-contained app folder notion is... difficult.  It could be done with a statement of minimum requirements, or inclusion of the library dependency chain outside of the minimum requirements, or a dependency check.  The dependency check itself is contextual... Hello World needs nothing; culture-sensitive sorting of a list needs ICU, etc.

For [Ubuntu 14.04.3](http://releases.ubuntu.com/14.04/ubuntu-14.04.3-desktop-amd64.manifest) it comes with libicu52, libcurl3, libssl1.0.0 (both libssl and libcrypto) and zlib1g... meaning all of the dependencies are there by default, but someone could still have removed them in trying to slim down their machine.

**Comment by davidfowl** [comment]
I don't know how you'd make any of that work in an app local world. The other platforms usually ship a runtime via the native package manager to avoid this problem. With an only nuget package xcopy the runtime world, you end up making this every app developers problem.

**Comment by piotrpMSFT** [comment]
Yep, agreed fully. @bartonjs @ellismg do you guys have enough knowledge to describe where we can run on stock OS images [sounds like Ubuntu 14.04.03 is one] where all of our known dependencies are in-box by default? How about Windows and OS X? It would be good to know where we 'just work' vs. where we 'work with caveats'.

@richlander @Petermarcu you guys might be interested in this issue.

**Comment by brthor** [comment]
Want to point out #293 here where we ran into the Ubuntu docker image not having libcurl.

**Comment by ellismg** [comment]
> do you guys have enough knowledge to describe where we can run on stock OS images [sounds like Ubuntu 14.04.03 is one] where all of our known dependencies are in-box by default?

What exactly do you mean by that?  For example, if the base install of Ubuntu 14.04 doesn't have the libunwind package installed, we aren't going to run, but it's just an `apt-get install` away.  Do you consider that "working with caveats" or "just works"?

For the binaries we produce for both CoreCLR and CoreFX we build against CentOS 7 and Ubuntu 14.04.  The CentOS 7 bits also run on RHEL 7 (since they share many of the same packages).  You may need to `apt-get` or `yum` dependencies.

If you are willing to build the native shims from source, there are many more options.  For example, I've built the native shims from source, across a bunch of different Ubuntu versions.

**Comment by bartonjs** [comment]
Oh, yeah, I didn't even think to check out the dependencies of libcoreclr.
```
jbarton@jsb-ubuntu2:~/coreclr$ ldd libcoreclr.so | cut -f2 --delimiter=\> | cut -f1 --delimiter=\( | xargs dpkg-query -S $@ | cut -f1 -d\ | sort -u
libc6:amd64:
libgcc1:amd64:
liblzma5:amd64:
libstdc++6:amd64:
libunwind8:
libuuid1:amd64:
```
Of those, libunwind8 is the only one that doesn't appear in the manifest for Ubuntu 14.04.03.

* OS X: Missing supported openssl (libssl/libcrypto > 1.0.0)
* Ubuntu 14.04.03: Missing libunwind8
* CentOS 7.1: Minimal install (the default that I did this morning): Missing libunwind, ICU
* Windows: Probably OK? :smile:

**Comment by brthor** [comment]
Created a blank azure vm, installed latest `dotnet-dev` package, and confirmed there are no missing dependencies.

@blackdwarf are you checking OSX?

**Comment by piotrpMSFT** [comment]
PKG does not offer as clean a solution as Apt-Get, but a warning would be nice.

@blackdwarf if we can show a warning that says there is a missing dependency, that would be great. Otherwise, let's add a note to the install sequence that states that the dependency must be installed.

**Comment by piotrpMSFT** [comment]
Fixed by #880
